Writing command line friendly applicaitons (#331)
gopheracademy/gopheracademy-web: content/advent-2019/cmdline.md

+++
title = "Writing Friendly Command Line Applications"

> date = "2019-12-08T00:00:00+00:00"

Reviewer: Please change the date to 2019-12-02 for publishing tomorrow

Author: Done

series = ["Advent 2019"]
author = ["Miki Tebeka"]
linktitle = "flag"
+++

Let me tell you a story...

In 1986 [Knuth](https://en.wikipedia.org/wiki/Donald_Knuth) wrote a program to
demonstrate [literate
programming](https://en.wikipedia.org/wiki/Literate_programming).

The task was to read a file of text, determine the n most frequently used
words, and print out a sorted list of those words along with their frequencies.
Knuth wrote a beautiful 10 page monolithic program.

Doug Mcllory read this and said 
`tr -cs A-Za-z '\n' | tr A-Z a-z | sort | uniq -c | sort -rn | sed ${1}q`

It's 2019, why am I telling you a story that happened 33 years ago? (Probably
before some of you were born). The computation landscape has changed a lot...
or has it?

The [Lindy effect](https://en.wikipedia.org/wiki/Lindy_effect) is a concept
that the future life expectancy of some non-perishable things like a technology
or an idea is proportional to their current age. TL;DR - old technologies are
here to stay.

If you don't believe me, see:

> - [oh-my-zsh](https://github.com/ohmyzsh/ohmyzsh) having close to 100,000 starts on GitHub

Reviewer 2: ```suggestion
- [oh-my-zsh](https://github.com/ohmyzsh/ohmyzsh) having close to 100,000 stars on GitHub
```

printing function which can be either text or JSON. Let's see an example:

```go
package main

import (
	"encoding/json"
	"flag"
	"fmt"
	"log"
	"os"
)

func main() {
	var jsonOut bool
	flag.BoolVar(&jsonOut, "json", false, "output in JSON format")
	flag.Parse()
	if flag.NArg() != 1 {
		log.Fatal("error: wrong number of arguments")
	}

	write := writeText
	if jsonOut {
		write = writeJSON
	}

	fi, err := os.Stat(flag.Arg(0))
	if err != nil {
		log.Fatalf("error: %s\n", err)
	}

	m := map[string]interface{}{
		"size":     fi.Size(),
		"dir":      fi.IsDir(),
		"modified": fi.ModTime(),
		"mode":     fi.Mode(),
	}
	write(m)
}

func writeText(m map[string]interface{}) {
	for k, v := range m {
		fmt.Printf("%s: %v\n", k, v)
	}
}

func writeJSON(m map[string]interface{}) {
	m["mode"] = m["mode"].(os.FileMode).String()
	json.NewEncoder(os.Stdout).Encode(m)
}

```

Then

```
$ ./finfo finfo.go
mode: -rw-r--r--
size: 783
dir: false
modified: 2019-11-27 11:49:03.280857863 +0200 IST
$ ./finfo -json finfo.go
{"dir":false,"mode":"-rw-r--r--","modified":"2019-11-27T11:49:03.280857863+02:00","size":783}
```

## Progress

Some operations can take long time, one way to make them faster is not by

> optimizing the code but by showing a spinner/progress bar. Don't beleive me,

Reviewer 2: ```suggestion
optimising the code but by showing a spinner/progress bar. Don't believe me,
```

> here's an excerpt from a [Nielsen

Reviewer 2: ```suggestion
here's an excerpt from [Nielsen
```

research](https://www.nngroup.com/articles/progress-indicators/)

> people who saw the moving feedback bar experienced higher satisfaction and
> were willing to wait on average 3 times longer than those who did not see any
> progress indicators.

### Spinner

Adding a spinner does not require any special packages:

```go
package main

import (
	"flag"
	"fmt"
	"os"
	"time"
)

var spinChars = `|/-\`

type Spinner struct {
	message string
	i       int
}

func NewSpinner(message string) *Spinner {
	return &Spinner{message: message}
}

func (s *Spinner) Tick() {
	fmt.Printf("%s %c \r", s.message, spinChars[s.i])
	s.i = (s.i + 1) % len(spinChars)
}

func isTTY() bool {
	fi, err := os.Stdout.Stat()
	if err != nil {
		return false
	}
	return fi.Mode()&os.ModeCharDevice != 0
}

func main() {
	flag.Parse()
	s := NewSpinner("working...")
	for i := 0; i < 100; i++ {
		if isTTY() {
			s.Tick()
		}
		time.Sleep(100 * time.Millisecond)
	}

}
```

Run it and you'll see a small spinner going.

### Progress Bar

> For progress bar, you'll probably need an external package such as

Reviewer 2: ```suggestion
For a progress bar, you'll probably need an external package such as
```
